scanpy/plotting/anndata.py

```diff
@@ -2000,3 +2239,154 @@ def translate_pos(pos_list, new_ticks, old_ticks):
+
+
+def _reduce_and_smooth(obs_tidy, goal_size):
+    """
+    Uses interpolation to reduce the number of observations (cells).
+    This is useful for plotting functions that otherwise will ignore
+    most of the cells' values.
+
+    The reduction and smoothing is only done per column
+
+    Parameters
+    ----------
+    obs_tidy : Pandas DataFrame. rows = obs (eg. cells), cols = vars (eg. genes)
+    goal_size : number of cells to keep
+
+    Returns
+    -------
+
+    """
+    if obs_tidy.shape[0] < goal_size:
+        return obs_tidy
+    else:
+        # usually, a large number of cells can not be plotted, thus
+        # it is useful to reduce the number of cells plotted while
+        # smoothing the values. This should be similar to an interpolation
+        # but done per row and not for the entire image.
+        from scipy.interpolate import UnivariateSpline
+        x = range(obs_tidy.shape[0])
+        # maximum number of cells to keep
+        new_x = np.linspace(0, len(x), num=goal_size, endpoint=False)
+        new_df = obs_tidy.iloc[new_x, :].copy()
+        for index, col in obs_tidy.iteritems():
+            spl = UnivariateSpline(x, col.values, s=20)
+            new_df[index] = spl(new_x)
+        return new_df.copy()
+
+
+def _plot_categories_as_colorblocks(groupby_ax, obs_tidy, colors=None, orientation='left', cmap_name='tab20'):
+    """
+    Plots categories as colored blocks. If orientation is 'left', the categories are plotted vertically, otherwise
+    they are plotted horizontally.
+
+    Parameters
+    ----------
+    groupby_ax : matplotlib ax
+    obs_tidy
+    colors : list of valid color names optional (default: `None`)
+        Color to use for each category.
+    orientation : `str`, optional (default: `left`)
+    cmap_name : `str`
+        Name of colormap to use, in case colors is None
+
+    Returns
+    -------
+    ticks position, labels, colormap
+    """
+
+    groupby = obs_tidy.index.name
+    from matplotlib.colors import LinearSegmentedColormap
+    if colors is None:
+        groupby_cmap = pl.get_cmap(cmap_name)
+    else:
+        groupby_cmap = LinearSegmentedColormap.from_list(groupby + '_cmap', colors, N=len(colors))
+
+    # determine groupby label positions such that they appear
+    # centered next/below to the color code rectangle assigned to the category
+    value_sum = 0
+    ticks = []  # list of centered position of the labels
+    labels = []
+    label2code = {}  # dictionary of numerical values asigned to each label
+    for code, (label, value) in enumerate(obs_tidy.index.value_counts(sort=False).iteritems()):
+        ticks.append(value_sum + (value / 2))
+        labels.append(label)
+        value_sum += value
+        label2code[label] = code
+
+    groupby_ax.grid(False)
+
+    if orientation == 'left':
+        groupby_ax.imshow(np.matrix([label2code[lab] for lab in obs_tidy.index]).T, aspect='auto', cmap=groupby_cmap)
```

Comment:
One thing I don’t get: The colors in the bar look pretty arbitrarily chosen from the colormap to me:

![grafik](https://user-images.githubusercontent.com/291575/48907731-3ba8f600-ee60-11e8-9b87-8e095f6ed764.png)

![grafik](https://user-images.githubusercontent.com/291575/48907740-406daa00-ee60-11e8-9c35-4c2c1ada6125.png)

but when i debug, the matrix you construct contains [0,0,0,0,…,1,1,1,1,…,2,2,2,…], so I’d expect the first colors to be chosen instead?

Reply:
ah, OK, fixed it in 426f028708cdd203b7d97d48eb558e695090da82